ml/nn/attention.go

```diff
@@ -0,0 +1,22 @@
+package nn
+
+import (
+	"github.com/ollama/ollama/ml"
+)
+
+func Attention(ctx ml.Context, query, key, value, mask ml.Tensor, scale float64) ml.Tensor {
+	if sdpa, ok := query.(ml.ScaledDotProductAttention); ok {
+		return sdpa.ScaledDotProductAttention(ctx, key, value, mask, scale)
```

Comment:
When will the query implement scaled dot product attention? Is this something that should handled here or in the caller?

Reply:
It's effectively a property of the backend since that's what defines the implementation of Tensor and whether there is a fused implementation of attention.

It should definitely be handled here - one of the main goals of this is to abstract away different implementations of attention from the model definition so we can drop in optimized versions.